File: python/2022/day01.py
def process(input: str) -> list[int]:
    return sorted([sum(map(int, r.splitlines())) for r in input.split("\n\n")], reverse=True)


def part1(input: str) -> int:
    return process(input)[0]


def part2(input: str) -> int:
    return sum(process(input)[:3])


TEST_INPUT = """1000
2000
3000

4000

5000
6000

7000
8000
9000

10000"""

PART1_TESTS = [
    (TEST_INPUT, 24000)
]

PART2_TESTS = [
    (TEST_INPUT, 45000)
]

File: rust/2022/src/bin/day01.rs
#[derive(Clone)]
struct Solution {
    reindeer: Vec<u32>,
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        let mut reindeer = Vec::new();
        let mut calories = 0;

        for line in input.lines() {
            if line.is_empty() {
                reindeer.push(calories);
                calories = 0;
            } else {
                calories += line.parse::<u32>()?;
            }
        }

        Ok(Self { reindeer })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        self.reindeer.iter().max().ok_or(anyhow!("empty input"))
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.reindeer.iter().take(3).sum::<u32>())
    }
}

aoc::solution!();
